service/history/workflow/workflow_task_state_machine.go

```diff
@@ -1087,11 +1080,64 @@ func (m *workflowTaskStateMachine) afterAddWorkflowTaskCompletedEvent(
 	attrs := event.GetWorkflowTaskCompletedEventAttributes()
 	m.ms.executionInfo.LastCompletedWorkflowTaskStartedEventId = attrs.GetStartedEventId()
 	m.ms.executionInfo.MostRecentWorkerVersionStamp = attrs.GetWorkerVersion()
-	if err := m.ms.CompleteDeploymentTransition(attrs.GetVersioningBehavior()); err != nil {
-		return err
+
+	wftDeployment := attrs.GetDeployment()
+	wftBehavior := attrs.GetVersioningBehavior()
+	versioningInfo := m.ms.GetExecutionInfo().GetVersioningInfo()
+
+	var completedTransition bool
+	if versioningInfo.DeploymentTransition != nil {
+		// It's possible that the completed WFT is not yet from the current transition because when
+		// the transition started, the current wft was already started. In this case, we allow the
+		// started wft to run and when completed, we create another wft immediately.
+		if versioningInfo.DeploymentTransition.GetDeployment().Equal(wftDeployment) {
+			versioningInfo.DeploymentTransition = nil
+			completedTransition = true
+		}
+	}
+
+	if m.ms.GetDeploymentTransition() != nil {
+		// There is still a transition going on. We need to schedule a new WFT so it goes to the
+		// transition deployment this time.
+		if _, err := m.ms.AddWorkflowTaskScheduledEvent(
+			false,
+			enumsspb.WORKFLOW_TASK_TYPE_NORMAL,
+		); err != nil {
+			return err
+		}
+	} else {
+		// Deployment and behavior before applying the data came from the completed wft.
+		wfDeploymentBefore := m.ms.GetEffectiveDeployment()
+		wfBehaviorBefore := m.ms.GetEffectiveVersioningBehavior()
+
+		// Change deployment and behavior based on the completed wft.
+		if wfBehaviorBefore != wftBehavior || !wfDeploymentBefore.Equal(wftDeployment) {
+			if versioningInfo == nil {
```

Comment:
It can't be `nil` here. At least I don't see how.

Reply:
It can: the wf was unversioned, hence its versioningInfo was nil. A speculative task redirected to a new deployment without touching versioningInfo at start time, and now it is completing with a non-nil wftDeployment.

Reply:
Line 1089 will give NPE then.